### ISSUE TITLE
RFC: #snapsafe: SysGenID - a system generation id provider

### ISSUE THREAD

**Issue by acatangiu** [systemd/systemd#19269]
**Is your feature request related to a problem? Please describe.**

Required in virtualized or containerized environments by applications that work with local copies or caches of world-unique data such as random values, uuids, monotonically increasing counters, cryptographic nonces, etc. Such applications can be negatively affected by VM or container snapshotting when the VM or container is either cloned or returned to an earlier point in time.

Solving the uniqueness problem strongly enough for cryptographic purposes requires a mechanism which can deterministically reseed userspace PRNGs with new entropy at restore time. This mechanism must also support the high-throughput and low-latency use-cases that led programmers to pick a userspace PRNG in the first place; be usable by both application code and libraries; allow transparent retrofitting behind existing popular PRNG interfaces without changing application code; it must be efficient, especially on snapshot restore; and be simple enough for wide adoption.

**Describe the solution you'd like**

Introduce a mechanism that standardizes an API for applications and libraries to be made aware of uniqueness breaking
events such as VM or container snapshotting, and allow them to react and adapt to such events.

The System Generation ID is meant to help in these scenarios by providing a monotonically increasing u32 counter that changes each time the VM or container is restored from a snapshot.

The `sysgenid` service exposes a monotonic incremental System Generation u32 counter via the DBus `com.RFC.sysgenid` accessible at `/com/RFC/sysgenid`. It provides asynchronous SysGen counter update notifications, as well as counter retrieval and confirmation mechanisms.
The counter starts from zero when the service is started and monotonically increments every time the system generation changes.

Userspace applications or libraries can (a)synchronously consume the system generation counter through the provided DBus interface, to make any necessary internal adjustments following a system generation update.

The provided DBus interface operations can be used to build a system level safe workflow that guest software can follow to protect itself from negative system snapshot effects.

System generation changes are driven by userspace software through a dedicated DBus method.

**Describe alternatives you've considered**
- System Generation ID kernel driver - https://lkml.org/lkml/2021/3/8/677
- System Generation ID as a userspace systemd plugin service (standalone service unit, not baked in systemd)

**The systemd version you checked that didn't have the feature you are asking for**
v248

**Comment by acatangiu** [comment]
Simple PoC implementation here: https://github.com/acatangiu/sysgenid-dbus

## SysGenID DBus interface

#### Terminology
 - `watcher` - a client using the SysGenID service _watching_ for system generation changes.
 - `untracked watcher` - default state for all clients. For a client to be tracked it has to explicitly opt-in by confirming back to the service the correct _system generation counter_.
 - `tracked watcher` - a client that is tracked by the service. Such a watcher is considered `up-to-date` only after confirming back to the service the correct _system generation counter_.
   Once tracked, a client is only _untracked_ when closing its connection to the DBus bus.
 - `outdated watcher` - a _tracked_ client that whose tracking has lived through a system generation change, but has not (yet) confirmed back to the service the correct _system generation counter_.

**Methods:**
- `GetSysGenCounter` - returns latest system generation counter.
- `AckWatcherCounter` - marks the client/watcher to be tracked for ACKs, is also used by the watcher to confirm/ack the correct _sys gen counter_ to the service after every generation change so the service keeps correct track of it as `outdated` or
  `up-to-date`.
  Will error if client/watcher confirms/acks the wrong _sys gen counter_.
- `CountOutdatedWatchers` - returns the number of current number of _outdated tracked watchers_.
  A value of `zero` can be interpreted as the system being fully re-adjusted after a generation change.
- `TriggerSysGenUpdate` - triggers a generation update (should be a privileged operation).

**Signals:**
- `NewSystemGeneration` - system generation change notification, also carries new _sys gen counter_.
- `SystemReady` - notification sent out when all tracked watchers have _acked_ the new _sys gen counter_. In other words, when all tracked software has adjusted to the new environment.

The service can keep track of watchers by DBus connections (`org.freedesktop.DBus.NameOwnerChanged`).

**Exported read-only file used for memory mappings:**

The service also exports the current _sys gen counter_ through a simple file.
The file contains only 4 bytes of data at offset 0, representing the u32 value of the system generation counter.
This file is meant to be mapped by other software in the system and be used as a low-latency generation counter probe mechanism in critical sections. This mmap() interface is targeted at libraries or code that needs to
check for generation changes in-line, where an event loop is not available or in cases where DBus calls are too expensive.
In such cases, logic can be added in-line with the sensitive code to check the counter and trigger on-demand/just-in-time readjustments when changes are detected on the memory mapped file.

Users of this interface that plan to lazily adjust most likely don't need to also use the DBus interface, since tracking or waiting on them doesn't make sense.

### Service interface DBus XML specification
```xml
<node name="/com/RFC/sysgenid">
  <interface name="com.RFC.sysgenid">
    <method name="AckWatcherCounter">
      <arg name="watcher_counter" type="u" direction="in"/>
      <arg name="sysgen_counter" type="u" direction="out"/>
    </method>
    <method name="CountOutdatedWatchers">
      <arg name="outdated_watchers" type="u" direction="out"/>
    </method>
    <method name="GetSysGenCounter">
      <arg name="sysgen_counter" type="u" direction="out"/>
    </method>
    <method name="TriggerSysGenUpdate">
      <arg name="min_gen" type="u" direction="in"/>
    </method>
    <signal name="NewSystemGeneration">
      <arg name="sysgen_counter" type="u"/>
    </signal>
    <signal name="SystemReady">
    </signal>
  </interface>
  <interface name="org.freedesktop.DBus.Introspectable">
    <method name="Introspect">
      <arg name="xml_data" type="s" direction="out"/>
    </method>
  </interface>
</node>
```

## Snapshot Safety Prerequisites and Example

If VM, container or other system-level snapshots happen asynchronously, at arbitrary times during an active workload there is no practical way to ensure that in-flight local copies or caches of world-unique data such as random values, secrets, UUIDs, etc are properly scrubbed and regenerated.
The challenge stems from the fact that the categorization of data as snapshot-sensitive is only known to the software working with it, and this software has no logical control over the moment in time when an external system snapshot occurs.

Let's take an OpenSSL session token for example. Even if the library code is made 100% snapshot-safe, meaning the library guarantees that the session token is unique (any snapshot that happened during the library call did not duplicate or leak the token), the token is still vulnerable to snapshot events while it transits the various layers of the library caller, then the various layers of the OS before leaving the system.

To catch a secret while it's in-flight, we'd have to validate system generation at every layer, every step of the way. Even if that would be deemed the right solution, it would be a long road and a whole universe to patch before we get there.

Bottom line is we don't have a way to track all of these in-flight secrets and dynamically scrub them from existence with snapshot
events happening arbitrarily.

### Simplifying assumption - safety prerequisite

**Control the snapshot flow**, disallow snapshots coming at arbitrary moments in the workload lifetime.

Use a system-level overseer entity that quiesces the system before snapshot, and post-snapshot-resume oversees that software components have readjusted to new environment, to the new generation. Only after, will the overseer un-quiesce the system and allow active workloads.

Software components can choose whether they want to be tracked and waited on by the overseer by using the marking themselves as tracked watchers.

The sysgenid service standardizes the API for system software to find out about needing to readjust and at the same time provides a mechanism for the overseer entity to wait for everyone to be done, the system to have readjusted, so it can un-quiesce.

### Example snapshot-safe workflow

1) Before taking a snapshot, quiesce the VM/container/system. Exactly how this is achieved is very workload-specific, but the general description is to get all software to an expected state where their event loops dry up and they are effectively quiesced.
2) Take snapshot.
3) Resume the VM/container/system from said snapshot.
4) Overseer will trigger generation bump using `TriggerSysGenUpdate` method.
5) Software components which have the DBus `NewGeneration` signal in their event loops are notified of the generation change. They do their specific internal adjustments. Some may have chosen to be tracked and waited on by the overseer, others might choose to do their adjustments out of band and not block the overseer. Tracked ones *must* signal when they are done/ready by confirming the new sys gen counter using the `AckWatcherCounter` DBus method.
6) Overseer will block and wait for all tracked watchers by waiting on the `SystemReady` DBus signal. Once all tracked watchers are done in step 5, the signal is sent by `sysgenid` service and overseer will know that the system has readjusted and is ready for active workload.
7) Overseer un-quiesces system.
8) There is a class of software, usually libraries, most notably PRNGs or SSLs, that don't fit the event-loop model and also have strict latency requirements. These can take advantage of the _exported read-only file used for memory mappings_. They can map the file and check sys gen counter value in-line with the critical section and can do so with low latency. When they are called after un-quiesce, they can just-in-time adjust based on the updated mapped value. For a well-designed service stack, these libraries should not be called while system is quiesced. When workload is resumed by the overseer, on the first call into these libs, they will safely JIT readjust. Users of this lazy on-demand readjustment model should not use the DBus interface or at least not enable watcher tracking since doing so would introduce a logical deadlock: lazy adjustments happen only after un-quiesce, but un-quiesce is blocked until all tracked watchers are up-to-date.

**Comment by poettering** [comment]
wouldn't a file in /run/ suffice for this, which is touched whenever the generation is bumped? i.e. why have an IPC interface for this?

i.e. have a file /run/generation/id or so that shall contain some randomized seed that the VM manager puts there through some agent or so, and refreshes on every new generation, ensuring that the mtime is updated when it does that. clients can then detect changes in four ways:

1. They can stat() the file and compare the mtime with the previous time they looked it. If it changed, the system generation was updated.
2. They can install an inotify watch on the inode of the file. They'll get a IN_CLOSE_WRITE event on it, whenever the generation changed, asynchronously.
3. They can read the file and use the contents as extra seed for their PRNG whenever they need a new random value
4. They can do the same via mmap(), so that they don't have to pay the cost for the syscalls each time.

It might even be OK to add a signal based API to PID 1 trigger the rollover. i.e. maybe by sending SIGRTMIN+7 to PID 1 it would update that seed file with data from the kernel's getrandom(). That way a VM agent would only have to enqueue SIGRTMIN+7 to PID 1 and that's it.

That all said, I am not sure this is really that great an approach, since this is all async, i.e. there will be a time window where the VM is already running again, where the consumers of this happily use the old PRNG seeds because the agent didn't ping PID 1 yet, or PID 1 didn't yet manage to update the id file.

Any truly race-free version would require some kernel support, i.e. some file in /proc or /sys or so that one can map and that is guaranteed to be already updated when the VM starts running again.

**Comment by poettering** [comment]
@bonzini @berrange sorry for the noise, but any idea if there's any infra worked on in this area anyway? what's your take on any of this? If you are not the right VM people to talk to about this, any chance you can CC the right people here? that'd be excellent. (you two i am aware work in the VM area, that's why I CCed you, sorry if you aren't the best people to comment on this)

**Comment by acatangiu** [comment]
Hi @poettering 

> wouldn't a file in /run/ suffice for this, which is touched whenever the generation is bumped? i.e. why have an IPC interface for this?
...
Any truly race-free version would require some kernel support, i.e. some file in /proc or /sys or so that one can map and that is guaranteed to be already updated when the VM starts running again.

When trying to support/fix this in kernel https://lkml.org/lkml/2021/2/9/828, we've seen that there is no truly race-free solution without userspace software cooperation.
Even if the kernel and/or hypervisor offer a race-free interface, there are still races within the userspace stack. A VM or container snapshot happening at arbitrary times can catch secrets (cryptographic nonces, session tokens, unique random values, etc) while they are in-flight within userspace stack (cached in some variable in code, in a socket buffer, etc). So even if we had a safe, race-free kernel mechanism that is used by, for example, a PRNG lib to make itself resilient to snapshot events, that'd only mean that the PRNG lib guarantees truly unique random value, but between the moment that unique random value leaves the PRNG lib and starts traversing the userspace stack to the moment it is ultimately consumed, it is again vulnerable to snapshot leaking or duplication...

As described in more detail in the `Snapshot Safety Prerequisites and Example` section from [here](https://github.com/systemd/systemd/issues/19269#issuecomment-816930693), one way to eliminate races is to introduce a system-level logical critical section where the system is quiesced (no in-flight secrets) and cooperates with userspace software to allow it to readjust to the new environment, before exiting this critical section and moving back to un-quiesced/active state.

The key here is to standardize this somehow and onboard libraries, runtimes, applications so that they can make themselves resilient to these snapshot events.
Current proposal is a transactional API, through which applications mark themselves as _ready_, to offer a system-level standard API that, if used by all components on the system, it would achieve `snapsafety`.

The actual IPC interface or system-level transactional API used is up for debate, but I don't think we can achieve #snapsafety without one.

**Comment by berrange** [comment]
> @bonzini @berrange sorry for the noise, but any idea if there's any infra worked on in this area anyway? what's your take on any of this? If you are not the right VM people to talk to about this, any chance you can CC the right people here? that'd be excellent. (you two i am aware work in the VM area, that's why I CCed you, sorry if you aren't the best people to comment on this)

On the hypervisor side, the conceptually related things I'm aware of are...

First Microsoft's  "Virtual Machine Generation ID" (http://go.microsoft.com/fwlink/?LinkId=260709). 

```The virtual machine generation ID is a feature whereby the virtual machines BIOS will 
expose a new ID. This is a 128-bit, cryptographically random integer value identifier that
 will be different every time the virtual machine executes from a different configuration
 file—such as executing from a recovered snapshot, or executing after restoring from
 backup. Programs running in a virtual machine can protect themselves from potential
 corruption by checking that the generation ID has not changed immediately prior to
 committing a transaction, they can also use the data provided in the 128-bit identifier
 as a high entropy random data source.```

This is implemented by HyperV first of all, and since by VMWare and QEMU/KVM (QEMU requires the mgmt app to actually opt-in and enable this feature)

This is essentially only usable as an after-the-fact notification of a snapshot that has already happened, but its better than nothing and given its implemented across these platforms, its probably the baseline you can expect. 

The SysGenID feature proposed here talks about a transactional model where there is synchronization before the snapshot is taken.  AFAICT, thats not viable with the VM Gen ID concept, and requires the hypervisor to interact with the guest OS more explicitly.

 In QEMU there is the qemu-guest-agent which has commands "fs-freeze" and "fs-thaw" which are sometimes used to quiesce the guest filesystem I/O before a snapshot and unquiesce it after the snapshot completes - not really needed for single disk VMs, but for multi-disk VMs they help get consistent snapshot state across all the disks. They are, however, general purpose commands that mgmt apps can use outside the context of snapshotting though, so it isn't something we could automatically tie to the SysGenID concept IMHO. They could potentially be enhanced though - eg they could take a optional parameter "bump-sysgenid: true" to let the host mgmt app indicate when they are being used in the context of of a snapshot op. If that's deemed inappropriate, then new QEMU guest agent commands would possibly be needed.

**Comment by acatangiu** [comment]
> This is implemented by HyperV first of all, and since by VMWare and QEMU/KVM (QEMU requires the mgmt app to actually opt-in and enable this feature)

As can be seen in the kernel patch thread https://lkml.org/lkml/2021/2/24/167 VmGenID is one option to drive SysGenID generations.

Unfortunately vmgenid device doesn't have an upstream driver yet. Once we settle on SysGenID design, we can also upstream a vmgenid driver.

**Comment by poettering** [comment]
Hmm, so in `systemd-logind` we have a concept of suspend and shutdown "inhibitors", that allow programs in a reasonably safe way to delay or block suspend and shutdown. Wouldn't it be sufficient to extend that to add a new kind of inhibitor lock "quiescevm" or so, which would work the same way, but could be used to delay quiescing?

Maybe we should even define a new system state you could enter with `systemctl quiesce` or so that would feel a lot like `systemctl suspend`, but instead of actually suspending we'd just tell some component that quescing is now complete. That way, a VM agent could do "systemctl quisce" and this would then start the inhibitor logic and run a few more services and we'd finally invoke some tool once everything is done, which by default would just be /bin/true or so. After it finishes we'd undo the quescing again for whatever that mean.

**Comment by acatangiu** [comment]
I did look at "inhibitors" and found them promising, but couldn't model these requirements on currently available inhibitors. Adding dedicated `quiesce` inhibitors should work beautifully!

**Comment by berrange** [comment]
> That way, a VM agent could do "systemctl quisce" and this would then start the inhibitor logic and run a few more services and we'd finally invoke some tool once everything is done, which by default would just be /bin/true or so. After it finishes we'd undo the quescing again for whatever that mean.

That's a bit of a challenging design from a virt host mgmt POV. We need the unquiesce to not happen until we've finished running operations on the host side. Actually it is worse, as sometimes the mgmt actions are not even invoked on the host, they're all triggered from some remote storage server. IOW, we can't just wait for some guest command to complete before unquiescing - we need to be able to explicitly control when it happens in some way that the mgmt app decides. This is why qemu splits its commands into a "freeze" and "thaw" pair

**Comment by poettering** [comment]
@berrange are there any reasonbly standardized APIs around for this?

The /bin/true thing was just a conceptual example: if a VM agent is installed it would of course not be /bin/true, but whatever thing you want it to be. Point was: that we go through the inhibitor dance, and then spawn a systemd transaction and at is its ultimate goal there'd be a binary we call, and the lifetime of that process defines the state where we are "quiescent". and then when the process exits we go back to normal.

The binary we call there would IRL probably just be a tool that does callouts, i.e. a tool that invokes all binaries that are in some directory, and then it waits synchronously for all of them to finish, and then exits. i.e. a VM agent would drop a binary in there, doing whatever if wants (could even just be a script). 

i.e. if you hack on an agent package you'd ship one of these dropins. And then whenever you want to queiesce the system you'd issue the equivalent of "systemctl quiesce" and wait until your binary is called. And then in that binary you can do whatever you want and once you exit from it the whole dance is reversed.

**Comment by acatangiu** [comment]
> i.e. if you hack on an agent package you'd ship one of these dropins. And then whenever you want to queiesce the system you'd issue the equivalent of "systemctl quiesce" and wait until your binary is called. And then in that binary you can do whatever you want and once you exit from it the whole dance is reversed.

Sounds valid to me, with the caveat that there needs to be a callback to VM agent at the end of the reversed dance as well.
VM agent needs to know when (within reasonable timeout) the system has finished adjusting to new environment so that it can unquiesce the VM/container (i.e. send traffic to it).

**Comment by berrange** [comment]
> i.e. if you hack on an agent package you'd ship one of these dropins. And then whenever you want to queiesce the system you'd issue the equivalent of "systemctl quiesce" and wait until your binary is called. And then in that binary you can do whatever you want and once you exit from it the whole dance is reversed.

So lets say we dropped a binary into /lib/systemd/quiesce/qemu-freeze

In the QEMU guest agent when getting ```fs-freeze``` command, we run ```systemctl quiesce``` and this triggers ```qemu-freeze```, which just goes into an infinite sleep.

When we get ```fs-thaw``` we can kill the running instance of ```qemu-freeze``` which lets the ```systemctl quiesce``` command complete.

That works when the QEMU guest agent triggers the quiesce, but this is an API that is exposed generally in systemctl to anyone/thing on the system.  If some other userspace service in the guest decides it needs to quiesce the system, it will trigger the ```qemu-freeze``` script that's in the declared systemd directory ```/lib/systemd/quiesce```, which will make the quiesce hang forever. 

IOW, I don't think we can use this scheme of dropping a script into a well defined directory because we need this to only run when it is the QEMU agent that's triggering the quiesce, not on any other triggers.

**Comment by poettering** [comment]
Well, I'd presume "qemu-freeze" would be a tiny bit smarter: if no "quiescing" was actually requested the tool would exit immediately to not hang things indefinitely. After all, one might install a multitude of agents into a single system, and each could drop a binary there. The binaries that notice that no request for quiescing was issued should all immediately exit, and only the ones where the circumstance indicate there's really reason to block should do so.

**Comment by poettering** [comment]
But, I'd like to step back in the discussion. i don't full grok why we'd need the "quiescing" thing. i.e. the original issue was about PRNG issues, i.e. cryptographic material. But for those an inhibitor lock sounds wrong to me. Wouldn't a much better pattern be there, that that kernel exposes that hyperv vmgenid memory buffer for everyone to mmap, and then openssl would mmap that and use it? Now of course, for long running operations that started before the system was duplicated and continues afterwards for a longer time one might think this might need a form of refresh notification. But does it really? wouldn't entirely suffice to check at reasonably times whether the vmgenid mmap is still the same as initially? and if not, abort things and start again? that sounds a lot simpler to me and would require zero further infra.

i.e. let's say you have a long running TLS connection going on, this is what would happen:

1. at the beginning of the TLS connection, openssl would seed its prng with the vmgenid, if it didn't already
2. it would then also copy vmgenid's current contents into a buffer associated with the TLS connection
3. right after submitting/accepting new TLS traffic it would check the current vmgenid. if still the same it would continue, otherwise immediately terminate the connection

i.e. by rechecking the vmgenid at the right places there would not be any need for notifications...

i.e what's the precise usecase for the quiescing here? you'd add non-trivial userspace infra for this, that requires buy-in from various projects to be useful. in particular any complex dbus api might have serious trouble of being acceptable to adoption by openssl for example.

Hence, my question: what usecase wouldn't sufficiently be covered f the vmgenid buffer would be exposed in procfs/sysfs as  mmapble thing thta can be check cheaply before *and* after each cryptographic operation that is it derived from it? Can you elaborate?

**Comment by acatangiu** [comment]
> i.e. by rechecking the vmgenid at the right places there would not be any need for notifications...
i.e what's the precise usecase for the quiescing here?
Hence, my question: what usecase wouldn't sufficiently be covered f the vmgenid buffer would be exposed in procfs/sysfs as mmapble thing thta can be check cheaply before and after each cryptographic operation that is it derived from it? Can you elaborate?

This is the solution we started from 😄 https://lkml.org/lkml/2020/10/20/315, but as we go deeper into the problem we see that needing **every** piece of software to do checks _after-the-fact_ at various steps during active workload wouldn't be practical.

Like you said, a practical solution that would better serve the industry is one that is applied to key components that serve entire classes of software - ideally with software built on top of them not having to do anything else.

i.e. one way is to have all of, for example, java or python libs and programs second-guess and double-check their transactions against a vm/system generation change, which is the original idea, but very hard to practically achieve;

the other way is to remove _the race_ in all the code out there with a solution that fixes it at the runtime level:
 1. _quiesce_ **the runtime** before the problem even appears (before snapshotting),
 2. do any necessary sanitization after snapshot while still quiescent - runtimes can even provide interface so existing code can do custom logic sanitization,
 3. once all these components (runtimes and all other system software that is registered as consumers of this mechanism) are done with sanitization, are readjusted to the new generation,
 4. _unquiesce_ and carry on.

> in particular any complex dbus api might have serious trouble of being acceptable to adoption by openssl for example

100% agree, not only because complex api but also because strict speed/latency requirements that OpenSSL has. That's why I believe in practice **we'd need to offer both**:
1. transactional API which delimits quiesced critical section for things like runtimes or any program with a main event loop,
2. something to `mmap()` that contains a sysgenid counter for example, which is going to go in the hotpath of libs like OpenSSL or PRNGs to be checked _before_ and _after_ all operations like you describe.

**Comment by poettering** [comment]
but what pecifically should the "quiesce" stuff be? Can you give more examples? Why do you think it's easier to add "quiescing" code paths to the many projects there are than it would be to add codepaths that recheck the vmgenid after-the-fact?

**Comment by agraf** [comment]
Let's maybe roll this backwards from the problems we're trying to solve. In all cases, we have a fully booted VM that gets snapshotted. After the snapshot, some data in the image is no longer unique, such as

  * PRNG seeds for anything using openssl
  * UUIDs in random code (sid in smbd or avahi for example)
  * Anything using /etc/machine-id

Why is that a problem? Usually it isn't at all, as long as systems are not connected. Only when network comes in, we need to make sure that uniqueness actually means something, because otherwise you can no longer identify all involved parties explicitly.

So what does quiescing look like effectively? IMHO all it takes is to make sure network is only available after the system has fully adapted to its new post-restore state. That means, we could do the following:

  * Block network before suspend (through an external helper), restore network only after all applications are adapted
  * Block network through hypervisor on resume. Notify HV through PV interface after all applications are adapted to unblock.
  * In serverless scenarios, only accept new requests after all applications are adapted

Library calls in code such as OpenSSL can adapt to the new world lazily by read()ing something on every callback they get. But that's a programming model that doesn't work well with most software out there. In smbd for example, I would ideally even want a script running that regenerates the SID and restarts the service. For <insert arbitrary self-built Java application here>, I would want an asynchronous code callback that allows me to reset internal variables within the right scope.

For that second case, we need something that is event driven, so that we can restore network access quickly. That's where mechanisms such as inotify unfortunately fall short.

Ideally, I would also love to just have /etc/machine-id. Then mmap() or for OpenSSL and use inotify for asynchronous event handling. But how do I know when it's safe to enable network again?

So IMHO the problem is not so much the quiescing, it's knowing when to end it. We can quiesce implicitly in the hypervisor just fine.

**Comment by poettering** [comment]
>     * Anything using /etc/machine-id

we document explicitly that machine IDs don't change during runtime, and remain stable during the entire runtime of the system from initrd to final shutdown. We do that explicitly to allow programs to cache it. Making this changable during runtime would break its documented semantics quite fundamentally.

**Comment by agraf** [comment]
That's the whole point of this effort. We're changing the semantics of the lifecycle of a system from "unique starting from the boot" to "can be cloned after boot".

If you prefer, we can also use a new file and slowly deprecate /etc/machine-id because its semantics no longer match reality in a snapshot driven world, but something has to change ;).

**Comment by poettering** [comment]
Humpf, this is an entirely different discussion then. And a really messy goal you have there. Quite frankly, I am not even sure what the machine ID is currently used for across the ecosystem. I know a lot of programs use it these days, it's basically everywhere and I have no idea how much of that can easily be reworked even to deal with changing machine ids (or alternatively a vmgen id concept).

**Comment by poettering** [comment]
I have the suspicion that if you want to do something like this, it'd be better to introduce this as new id, and leave the machine id the way it is, and then convince people to use the new id preferably if it exists, and fall back to the machine ID if it doesn't.

The microsoft vmgenid stuff, how is that supposed to be used? as a change detector only (i.e. if vmgenid changed we know we are in a new instance)? as an additional seed for a prng (if so, this would mean it would need to stay secret, but that would defeat the whole purpose)? As both a change detector and a machine-id-like ID that is accessible by local processes easily?

Assuming the answer is the latter, I think it might make sense to do it like this:

1. expose the vmgenid in a 128bit hex value (maybe uuid format) in /sys or /proc somewhere, world-readable, either matching the format of `/proc/sys/kernel/random/boot_id` or in `/etc/machine-id`, you pick.
2. introduce a new file /run/sysgen-id or so that is either a symlink to the aforementioned kernel-provided file, or is managed in userspace by PID 1 (for container mgrs, where the generation should expose the snapshot logic of the container, not of the surrounding host)
3. declare that inotify, mmap, mtime all work on the file as one would expect to, regardless if kernel provided or userspace provided

Then leave /etc/machine-id the way it is (i.e. make it the "origin" ID if you so will, i.e. the ID the whole tree of system "branches" has, and make /run/sysgen-id the ID for each "branch" of the tree). next, teach code all over the place to use /run/sysgen-id if they are looking for an ID that changes for new instances of the same snapshotted system, and ideally use mmap/mtime/inotify/content checks to detect changes. Tell them to use /etc/machine-id as fallback.

And then, after that#s done add some new concept for doing "new-generation-id" event handling. As I understand now, this should be something where some daemons (networkd, NetworkManager, …) can acquire some objects as party that is notified, and other daemons can acquire some objects as party that is notifying (i.e. anything openssl-using, avahi, …). When the system is about to generate a new ID, it will first notify the first kind of daemons that this is going to happen, which will make them shut off all networking. This could be called "quiescing". Now, the system is duplicated, and when coming back the system will then notify the second kind of programs, that this happened and give them some time to react to that. (this could be called un-quiescing) Once all reacted (or a timeout passed) it will let the first kind of daemons know this, so that they turn on networking again.

I doubt this would fit into logind's inhibitors concept. Also: yuck, for the complexity.

What's your intention with the boot id btw? Do you intend to update `/proc/sys/kernel/random/boot_id` on each new genration (i'd really not).

**Comment by poettering** [comment]
> I doubt this would fit into logind's inhibitors concept. 

Or maybe it actually would fit into it. there'd be two actions that can be inhibited: "quiesce" and "unquiesce". networkd/NetworkManager would install a delay inhibitor for the former, and openssl/avahi/… for the latter. Then, when the dance is started we first do the "queiscing" inhibitor dance. Then we are officially "quiesced". Then we'll do the "unquiescing" dance, and once that's achieved we are back running again.

So the system could have four states: running → queiscing → quiesced → unquiescing → running

**Comment by acatangiu** [comment]
Can these inhibitors be (easily/safely) exposed to unprivileged containers?

**Comment by acatangiu** [comment]
answering some earlier questions:
> The microsoft vmgenid stuff, how is that supposed to be used? as a change detector only (i.e. if vmgenid changed we know we are in a new instance)? as an additional seed for a prng (if so, this would mean it would need to stay secret, but that would defeat the whole purpose)?

There is no upstream linux driver available yet so it can be either (but not both). I am leaning on using the actual 128bit id as additional kernel entropy (and id never leaves the kernel) and only expose a simple counter to userspace to be used as change detector.

> What's your intention with the boot id btw? Do you intend to update /proc/sys/kernel/random/boot_id on each new genration (i'd really not).

I would not either, `boot_id` is explicitly stated to change only as a result of (re)boot. There will be applications however that are currently using `boot_id` and which would have to move to using whichever new id file we will expose for snap events.

**Comment by bluca** [comment]
> Can these inhibitors be (easily/safely) exposed to unprivileged containers?

Existing inhibitors were designed and implemented specifically to be used by unprivileged users (eg: click reboot button on gui), and then expanded to other usage, so I'd say yes

**Comment by acatangiu** [comment]
> Can these inhibitors be (easily/safely) exposed to unprivileged containers?

I see that inhibitors are exposed by the `org.freedesktop.login1` [1] DBus service. Is this service exposed in containers (docker, podman, etc)?

[1] - https://github.com/systemd/systemd/blob/main/src/login/inhibit.c

**Comment by bluca** [comment]
> > Can these inhibitors be (easily/safely) exposed to unprivileged containers?
> 
> I see that inhibitors are exposed by the `org.freedesktop.login1` [1] DBus service. Is this service exposed in containers (docker, podman, etc)?
> 
> [1] - https://github.com/systemd/systemd/blob/main/src/login/inhibit.c

As far as I know either the system bus is exposed, by binding the dbus socket, or it is not - it's not a per-service thing. Authorization thorough polkit will be per-service, but that's no different from accessing from the host.

**Comment by acatangiu** [comment]
@poettering
Apologies for the absence on this thread. We, in AWS, have thought about this some more and also modeled some service components using the approaches discussed on this thread. The results are encouraging, and this looks like a good place to start adding #snapsafe patches to the ecosystem.

The path we had most success with can be best described here by starting from your previous proposal (as it’s largely based on that):

> I have the suspicion that if you want to do something like this, it'd be better to introduce this as new id, and leave the machine id the way it is, and then convince people to use the new id preferably if it exists, and fall back to the machine ID if it doesn't.

Agreed.

> The microsoft vmgenid stuff, how is that supposed to be used? as a change detector only (i.e. if vmgenid changed we know we are in a new instance)? as an additional seed for a prng (if so, this would mean it would need to stay secret, but that would defeat the whole purpose)? As both a change detector and a machine-id-like ID that is accessible by local processes easily?

There is no upstream linux driver available yet so it can be either (but not both). I am leaning on using the actual 128bit id as additional kernel entropy (value never leaves the kernel) and only expose a simple counter to userspace to be used as change detector; u32 counter can also be atomically read so it makes everything easier.

> Assuming the answer is the latter, I think it might make sense to do it like this:
> expose the vmgenid in a 128bit hex value (maybe uuid format) in /sys or /proc somewhere, world-readable, either matching the format of /proc/sys/kernel/random/boot_id or in /etc/machine-id, you pick.

Yes, but u32 counter instead of 128bit UUID.

> introduce a new file /run/sysgen-id or so that is either a symlink to the aforementioned kernel-provided file, or is managed in userspace by PID 1 (for container mgrs, where the generation should expose the snapshot logic of the container, not of the surrounding host)
> declare that inotify, mmap, mtime all work on the file as one would expect to, regardless if kernel provided or userspace provided

Agreed.

This model is also backwards compatible while being much more flexible with the kernel SysGenID patch. For example, there’s a PR (https://github.com/awslabs/aws-lc/pull/179) [1] in AWS libcrypto (AWS crypto lib based on BoringSSL and OpenSSL) that makes AWS-LC #snapsafe using kernel SysGenID device. For this lib and other *SSL libs, the same patch/change will work out of the box with approach described above - only thing different would be the configurable file path: /run/sysgen-id instead of /dev/sysgenid.

Then there’s the second half of the story, for those components that need to do "new-generation-id" event coordinating.
For this, we modeled and validated the proposed quiesce/snapshot and unquiesce/resume inhibitors:

> Or maybe it actually would fit into it. there'd be two actions that can be inhibited: "quiesce" and "unquiesce". networkd/NetworkManager would install a delay inhibitor for the former, and openssl/avahi/… for the latter. Then, when the dance is started we first do the "queiscing" inhibitor dance. Then we are officially "quiesced". Then we'll do the "unquiescing" dance, and once that's achieved we are back running again. 
> So the system could have four states: running → queiscing → quiesced → unquiescing → running

For example, these inhibitors can be used as a backend for [Java Coordinated Restore at Checkpoint](https://crac.github.io/jdk/jdk-crac/api/java.base/jdk/crac/package-summary.html) [2] as a possible mechanism to make Java Applications #snapsafe.

I strongly believe if we provide the above in systemd, we would provide a great foundational step to fuel further innovation in this space.

How should we move forward? Would you accept the above if we were to contribute it to systemd?

[1] https://github.com/awslabs/aws-lc/pull/179
[2] https://crac.github.io/jdk/jdk-crac/api/java.base/jdk/crac/package-summary.html

**Comment by acatangiu** [comment]
Closing this in favor of #20222.

**Comment by rwmjones** [comment]
Just a note that in addition to VMware and QEMU, Hyper-V also supports the 128 bit VM Generation ID (really UUID).  Also libvirt supports it.
